platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/report/generic/PerRootTabRenderer.java

```diff
@@ -83,11 +88,50 @@ protected void render(TestTreeModel.PerRootInfo info, SimpleHtmlWriter htmlWrite
             if (info.getChildren().isEmpty()) {
                 renderLeafDetails(info, htmlWriter);
             } else {
-                renderChildren(htmlWriter);
+                renderChildTabs(htmlWriter);
             }
             htmlWriter.endElement();
         }
 
+        private void renderChildTabs(SimpleHtmlWriter htmlWriter) throws IOException {
+            List<Pair<Path, TestTreeModel.PerRootInfo>> children = Streams.of(getCurrentModel().getChildrenOf(rootIndex))
+                .flatMap(t ->
+                    t.getPerRootInfo().get(rootIndex).stream()
+                        .map(p -> Pair.of(t.getPath(), p))
+                )
+                .collect(Collectors.toList());
+            List<Pair<String, ChildListRenderer>> childListRenderers = new ArrayList<>();
+
+            addResultTabIfNeeded("Failed", TestResult.ResultType.FAILURE, children, childListRenderers);
+            addResultTabIfNeeded("Skipped", TestResult.ResultType.SKIPPED, children, childListRenderers);
+            childListRenderers.add(Pair.of("All", new ChildListRenderer(children)));
+
+            ReportRenderer<TestTreeModel, SimpleHtmlWriter> renderer;
+            if (childListRenderers.size() == 1) {
+                // No need to render tabs if there is only one tab
+                renderer = Objects.requireNonNull(childListRenderers.get(0).right);
+            } else {
+                TabsRenderer<TestTreeModel> tabsRenderer = new TabsRenderer<>();
+                childListRenderers.forEach(p -> tabsRenderer.add(p.left, p.right));
+                renderer = tabsRenderer;
+            }
+            renderer.render(getCurrentModel(), htmlWriter);
```

Comment:
🤔 I think it would be clearer to split this method.  It's named `renderChildTabs`, but it sometimes doesn't result in child tabs being rendered.  It would be confusing to debug a report that you can see lacks tabs and find yourself stepping through this method.

Maybe something like:

```renderNonLeafSummary() {
	List renderers = getChildTabRenderers(); // lines 97-107
	... // continue with line 110
}```